app/upgrades/v17/upgrades.go

```diff
@@ -124,12 +89,225 @@ func CreateUpgradeHandler(
+func mainnetUpgradeHandler(ctx sdk.Context, keepers *keepers.AppKeepers, communityPoolAddress sdk.AccAddress, poolLinks *[]gammmigration.BalancerToConcentratedPoolLink, fullRangeCoinsUsed *sdk.Coins) error {
+	assetPairs := InitializeAssetPairs(ctx, keepers)
+
+	for _, assetPair := range assetPairs {
+		clPoolDenom, clPoolId, err := createCLPoolWithCommunityPoolPosition(ctx, keepers, assetPair.LinkedClassicPool, assetPair.BaseAsset, assetPair.SpreadFactor, communityPoolAddress, poolLinks, fullRangeCoinsUsed)
+		if err != nil {
+			return err
+		}
+
+		err = authorizeSuperfluidIfEnabled(ctx, keepers, assetPair.LinkedClassicPool, clPoolDenom)
+		if err != nil {
+			return err
+		}
+
+		err = manuallySetTWAPRecords(ctx, keepers, clPoolId)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// testnetUpgradeHandler creates CL pools for all existing balancer pools. It also links the CL pools to their balancer pool counterpart, creates a full range position with the community pool,
+// authorizes superfluid for the CL pool if the balancer pool is superfluid enabled, and manually sets the TWAP records for the CL pool.
+func testnetUpgradeHandler(ctx sdk.Context, keepers *keepers.AppKeepers, communityPoolAddress sdk.AccAddress, poolLinks *[]gammmigration.BalancerToConcentratedPoolLink, fullRangeCoinsUsed *sdk.Coins) error {
+	// Retrieve all GAMM pools on the testnet.
+	pools, err := keepers.GAMMKeeper.GetPools(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, pool := range pools {
+		skipPool, gammPoolId, baseAsset, spreadFactor, err := testnetParsePoolRecord(ctx, pool, keepers)
+		if err != nil {
+			return err
+		}
+		if skipPool {
+			continue
+		}
+
+		clPoolDenom, clPoolId, err := createCLPoolWithCommunityPoolPosition(ctx, keepers, gammPoolId, baseAsset, spreadFactor, communityPoolAddress, poolLinks, fullRangeCoinsUsed)
+		if errors.Is(err, notEnoughLiquidityForSwapErr) {
+			continue
+		} else if err != nil {
+			return err
+		}
+
+		err = authorizeSuperfluidIfEnabled(ctx, keepers, gammPoolId, clPoolDenom)
+		if err != nil {
+			return err
+		}
+
+		err = manuallySetTWAPRecords(ctx, keepers, clPoolId)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// createCLPoolWithCommunityPoolPosition creates a CL pool for a given balancer pool and adds a full range position with the community pool.
+func createCLPoolWithCommunityPoolPosition(ctx sdk.Context, keepers *keepers.AppKeepers, gammPoolId uint64, baseAsset string, spreadFactor sdk.Dec, communityPoolAddress sdk.AccAddress, poolLinks *[]gammmigration.BalancerToConcentratedPoolLink, fullRangeCoinsUsed *sdk.Coins) (clPoolDenom string, clPoolId uint64, err error) {
+	// Check if classic pool has enough liquidity to support a 0.1 OSMO swap before creating a CL pool.
+	// If not, skip the pool.
+	osmoIn := sdk.NewCoin(QuoteAsset, sdk.NewInt(100000))
```

Comment:
Why is 0.1 OSMO chosen and why does it work for all pools?

Reply:
We just need some dust amount to create the cl/pool/x denom. 0.1 is not a magic number, its just some value used to swap in the gamm pool to get the baseDenom out, so the community pool can create a full range position with both assets.